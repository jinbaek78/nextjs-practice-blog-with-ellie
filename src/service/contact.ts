import { EmailData } from './email';

export async function sendContactEmail(email: EmailData) {
  const response = await fetch('http://localhost:3000/api/contact', {
    method: 'POST',
    body: JSON.stringify(email),
    headers: {
      'Content-Type': 'application/json',
    },
  });

  const data = await response.json();
  if (!response.ok) {
    throw new Error(data.message || 'failed to request to server');
  }

  return data;
}
