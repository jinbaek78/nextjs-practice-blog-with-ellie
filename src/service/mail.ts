import { BannerData } from '@/components/Banner';
import { Form } from '@/components/ContactForm';
import { object, string, number, date, InferType } from 'yup';

export async function sendEmail(
  data: Form
  // callback: (banner: BannerData) => void
): Promise<BannerData> {
  let formSchema = object({
    subject: string().required(),
    from: string().email().required(),
    message: string().min(2).required(),
  });

  let dataValidationResult;
  try {
    dataValidationResult = await formSchema.validate(data);
  } catch (err) {
    console.log(err);
    return { message: 'please, write the right format', state: 'error' };
  }

  const formHeader = new Headers();
  formHeader.append('Content-Type', 'application/json');

  const requestOptions = {
    method: 'POST',
    headers: formHeader,
    body: JSON.stringify(data),
  };

  let res;
  let resultBanner: BannerData;
  try {
    res = await fetch('http://localhost:3000/api/contact', requestOptions);
  } catch (err) {
    console.log(err);
    return (resultBanner = { message: 'fail', state: 'error' });
  }

  return res.status === 200
    ? { message: 'success', state: 'success' }
    : { message: 'fail', state: 'error' };

  // return fetch('http://localhost:3000/api/contact', requestOptions)
  //   .then((response) => response.json())
  //   .then((result) => console.log(result))
  //   .catch((error) => console.log('error', error));
}
