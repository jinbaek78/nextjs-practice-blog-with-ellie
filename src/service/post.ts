import fs from 'fs/promises';
import path from 'path';
import { cwd } from 'process';

export type Post = {
  title: string;
  description: string;
  date: string;
  category: string;
  path: string;
  featured: boolean;
};

export async function getPostsData(): Promise<Post[]> {
  const postsData = await JSON.parse(
    await fs.readFile(`${path.join(cwd(), 'public', 'data', 'posts.json')}`, {
      encoding: 'utf-8',
    })
  );

  return postsData;
}
