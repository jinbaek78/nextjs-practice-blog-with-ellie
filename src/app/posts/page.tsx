import Posts from '@/components/Posts';
import { getAllPosts } from '@/service/posts';

export default async function PostsPage() {
  const posts = await getAllPosts();

  return <Posts posts={posts} />;
}
