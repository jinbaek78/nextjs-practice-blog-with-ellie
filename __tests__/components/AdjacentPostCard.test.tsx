import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';
import mockRouter from 'next-router-mock';
import { MemoryRouterProvider } from 'next-router-mock/MemoryRouterProvider';
import postData from '../../public/data/posts.json';
import { Post } from '@/service/posts';
import AdjacentPostCard from '@/components/AdjacentPostCard';
import { FaArrowLeft, FaArrowRight } from 'react-icons/fa';

jest.mock('react-icons/fa', () => ({
  FaArrowLeft: jest.fn(() => console.log('fake left called')),
  FaArrowRight: jest.fn(() => console.log('fake right called')),
}));

describe.skip('AdjacentPostCard', () => {
  const posts: Post[] = postData;
  const postInfo = posts[0];
  const { path, title, description } = postInfo;

  afterEach(() => {
    (FaArrowLeft as jest.Mock).mockReset();
    (FaArrowRight as jest.Mock).mockReset();
  });

  xit('renders correctly', () => {
    const { container } = render(
      <AdjacentPostCard post={postInfo} type="prev" />
    );
    expect(container).toMatchSnapshot();
  });

  xit('renders with props correctly', () => {
    const postInfo = posts[0];
    render(<AdjacentPostCard post={postInfo} type="prev" />);
    const link = screen.getByRole('link') as HTMLLinkElement;

    expect(screen.getByText(title)).toBeInTheDocument();
    expect(screen.getByText(description)).toBeInTheDocument();
    expect(link.href).toBe(`http://localhost/posts/${path}`);
  });

  it('navigates to the clicked post page when the previous post is clicked', async () => {
    render(<AdjacentPostCard post={postInfo} type="prev" />, {
      wrapper: MemoryRouterProvider,
    });
    const button = screen.getByRole('link');
    await fireEvent.click(button);
    expect(mockRouter.asPath).toEqual(`/posts/${path}`);
    expect(FaArrowLeft).toHaveBeenCalledTimes(1);
    expect(FaArrowRight).not.toBeCalled();
  });

  it('navigates to the clicked post page when the next post is clicked', async () => {
    const nextPost = posts[2];
    const { path, title, description } = nextPost;
    render(<AdjacentPostCard post={nextPost} type="next" />, {
      wrapper: MemoryRouterProvider,
    });
    const button = screen.getByRole('link');
    await fireEvent.click(button);
    expect(mockRouter.asPath).toEqual(`/posts/${path}`);
    expect(FaArrowRight).toHaveBeenCalledTimes(1);
    expect(FaArrowLeft).not.toBeCalled();
  });
});
