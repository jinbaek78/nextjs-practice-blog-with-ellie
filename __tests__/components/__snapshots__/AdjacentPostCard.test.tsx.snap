// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdjacentPostCard renders correctly 1`] = `
<div>
  <a
    class="group relative w-full max-h-56 bg-black"
    href="/posts/review-2021"
  >
    <img
      alt="2021년 리뷰"
      class="w-full opacity-40"
      data-nimg="1"
      decoding="async"
      height="100"
      loading="lazy"
      src="/_next/image?url=%2Fimages%2Fposts%2Freview-2021.png&w=384&q=75"
      srcset="/_next/image?url=%2Fimages%2Fposts%2Freview-2021.png&w=256&q=75 1x, /_next/image?url=%2Fimages%2Fposts%2Freview-2021.png&w=384&q=75 2x"
      style="color: transparent;"
      width="150"
    />
    <div
      class="absolute top-1/2 left-1/2 -translate-x-1/2 -translate-y-1/2 w-full flex justify-around items-center  text-white"
    >
      <svg
        class="text-5xl m-4 text-yellow-300 transition-all group-hover:text-6xl mx-10"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 448 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M257.5 445.1l-22.2 22.2c-9.4 9.4-24.6 9.4-33.9 0L7 273c-9.4-9.4-9.4-24.6 0-33.9L201.4 44.7c9.4-9.4 24.6-9.4 33.9 0l22.2 22.2c9.5 9.5 9.3 25-.4 34.3L136.6 216H424c13.3 0 24 10.7 24 24v32c0 13.3-10.7 24-24 24H136.6l120.5 114.8c9.8 9.3 10 24.8.4 34.3z"
        />
      </svg>
      <div
        class="w-full text-center"
      >
        <h3
          class="text-3xl font-bold"
        >
          2021년 리뷰
        </h3>
        <p
          class="font-bold"
        >
          2021년 한해 동안 배운 기술들, 생각들 정리
        </p>
      </div>
    </div>
  </a>
</div>
`;
